added Authentication Guide and Supabase Configuration with Astro to supabase.mdx (#4371)
withastro/docs: src/content/docs/en/guides/backend/supabase.mdx

description: Add a backend to your project with Supabase
type: backend
service: Supabase
stub: true
stub: false
---
import FileTree from '~/components/FileTree.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

[Supabase](https://supabase.com/) is an open source Firebase alternative. It provides a Postgres database, authentication, edge functions, realtime subscriptions, and storage.

> ## Community Resources

Reviewer: We should keep the community resources at the end of the guide

- .env
- ...
</FileTree>

### Install Dependencies

To connect Astro with Supabase, install the following package using the command below:

- `@supabase/supabase-js`

<PackageManagerTabs>
  <Fragment slot="npm">
    ```shell
    npm install @supabase/supabase-js
    ```
  </Fragment>
  <Fragment slot="pnpm">
    ```shell
    pnpm install @supabase/supabase-js
    ```
  </Fragment>
  <Fragment slot="yarn">
    ```shell
    yarn add @supabase/supabase-js
    ```
  </Fragment>
</PackageManagerTabs>

### Create Profile Table

1. Go to [Studio URL](http://localhost:54323)
2. Click on SQL Editor on the left sidebar
3. Choose `User Management Starter` from `Quick start` in the bottom of the page
4. Click `RUN`

### Create Supabase Client

1. Create a directory named `lib` in the `src/` directory.
2. Create a file named `supabaseClient.ts` in the `src/lib/` directory.

In `supabaseClient.ts` add the following code to create the supabase client.

```ts title="src/lib/supabaseClient.ts"
import { createClient } from '@supabase/supabase-js'

export const supabase = createClient(
  `${import.meta.env.SUPABASE_URL}`,
  `${import.meta.env.SUPABASE_ANON_KEY}`,
  {         
    auth: {

>       persistSession: false,

Reviewer: Since authentication will be handled server side we should enable `pkce`

Author: Exactly! I will fix it soon.

    }
  }
)
```

The project structure should include these new files:

<FileTree title="Project Structure">
- ...
- src/
  - lib/
    - supabaseClient.ts
  - env.d.ts
- .env
- astro.config.mjs
- package.json
- ...
</FileTree>

### Prepare Authentication Functions

- Create a file named `auth.ts` in the `src/lib/` directory.

In `auth.ts` add the following code to create the authentication functions:

> ```ts title="src/lib/auth.ts"

Reviewer: As a good practice, we should try to minimize the use utility functions. Even though we will repeat a lot of code, it helps users see which functions we are using from the original library.

import { supabase } from "./supabaseClient";

export async function signUp(email: string, password: string) {
    return await supabase.auth.signUp({
        email,
        password,
        options: {
          emailRedirectTo: 'http://localhost:3000/profile'
        }
    })
}

export async function signIn(email: string, password: string) {
    return await supabase.auth.signInWithPassword({
        email,
        password
    })
}

export async function signInWithGithub() {
    return await supabase.auth.signInWithOAuth({
        provider: 'github',
        options: {
          redirectTo: 'http://localhost:3000/profile'
        }
    });
}

export async function getSessionData() {
    return await supabase.auth.getSession();
}

export async function setSessionData(access_token: string, refresh_token: string) {
    return await supabase.auth.setSession({
        access_token,
        refresh_token,
    })
}

export function AuthState() {
    return supabase.auth.onAuthStateChange((event, session) => {
        console.log(event, session);
    })
}

export async function signOut() {
    return await supabase.auth.signOut()
}
```

:::tip

> Check [Supabase User Management](https://supabase.com/dashboard/project/qiilhzulchhvmptaayfz/api?page=users) to learn more about these functions.

Reviewer: I think this link is for your specific project, I don't think it will work for everyone.

Author: > I think this link is for your specific project, I don't think it will work for everyone.

Fixed it. Thank you :heart:

                    aria-live="polite"
                    type="submit"
                >
                    Sign In
                </button>
                <p class="text-center text-gray-800">Need an account? <a href="/signup">Sign Up</a></p>
            </form>
        </div>
    </div>        
</Layout>

<script>
    import { signIn, getSessionData } from "../lib/auth";

    const form = document.querySelector("form");

    form?.addEventListener("submit", async (e) => {
        e.preventDefault();
        const formData = new FormData(e.target as HTMLFormElement);
        const creds = {
            email: formData.get("email") as string,
            password: formData.get("password") as string,
        };

        const signin = await signIn(creds.email, creds.password);
        if (signin.error !== null) {
            console.log(signin.error);
            alert(signin.error.message);
            window.location.href = "/signin";
        } else {
            const session = await getSessionData();

            console.log(session);

            const maxAge = 15 * 24 * 60 * 60;
            if(session) {

>                 document.cookie = `sb-access-token=${session.data.session?.access_token}; path=/; expires=${session.data.session?.expires_at}; max-age=${maxAge}; SameSite=Lax; secure`;

Reviewer: Astro has utility functions to work with cookies. We could replace `document.cookie` with `Astro.cookies`

            alert("Passwords do not match");
            return;
        } else {
            try {
                await signUp(creds.email, creds.password);
            } catch (error) {
                console.log(error);
            }
        }
    });

    const github = document.getElementById("github");
    github?.addEventListener("click", async (e) => {
        e.preventDefault();
        await signInWithGithub();
    })
</script>
```

### Profile

1. Go to `src/pages`
2. Create a new file named `profile.astro`
3. Add the following code to `profile.astro`:

```astro
---
// src/pages/profile.astro
import Layout from "../layouts/Layout.astro";
import { getSessionData, setSessionData, AuthState} from "../lib/auth";

let accessToken = Astro.cookies.get("sb-access-token").value
let refreshToken = Astro.cookies.get("sb-refresh-token").value

console.log("accessToken", accessToken);
console.log("refreshToken", refreshToken);
const session = await setSessionData(accessToken, refreshToken);

console.log("session", session);

if (session.data.session === null || (accessToken === undefined && refreshToken === undefined)) {
	return Astro.redirect("/signin");
}

> console.log("Auth State", AuthState())

Reviewer: I don't think we need to set up auth state since the session management is handled server side.

Author: Yes I figured it out later. I will fix it and update this guide soon.

description: Add a backend to your project with Supabase
type: backend
service: Supabase
stub: false
---
import FileTree from '~/components/FileTree.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

[Supabase](https://supabase.com/) is an open source Firebase alternative. It provides a Postgres database, authentication, edge functions, realtime subscriptions, and storage.

## Getting Started

### Prerequisites
- [Supabase CLI](https://supabase.com/docs/guides/cli/getting-started) for local development.
- An Astro project with [server-side rendering (SSR)](/en/guides/server-side-rendering/#enabling-ssr-in-your-project) enabled.

> ### Initialize Supabase
> Run `supabase init` in the root of the project, it should include this folder:

Reviewer 2: ```suggestion
### Initialize Supabase

Run `supabase init` in the root of the Astro project, it should create this folder:
```

<FileTree title="Project Structure">
- ...
- supabase/
  - config.toml
  - functions/
  - migrations/
  - seed.sql
- ...
</FileTree>

> ### Start Supabase
> Run `supabase start` to start supabase stack. it should return this:

Reviewer 2: ```suggestion
### Start Supabase

Run `supabase start` to start Supabase stack. It should return this:
```

> ```ini title="Terminal"
> Started supabase local development setup.
> 
>          API URL: http://localhost:54321
>      GraphQL URL: http://localhost:54321/graphql/v1
>           DB URL: postgresql://postgres:postgres@localhost:54322/postgres
>       Studio URL: http://localhost:54323
>     Inbucket URL: http://localhost:54324
>       JWT secret: super-secret-jwt-token-with-at-least-32-characters-long
>         anon key: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0
> service_role key: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0.EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU
> ```

Reviewer 2: ``````suggestion

```ini title="Terminal"
Started supabase local development setup.

         API URL: http://localhost:54321
     GraphQL URL: http://localhost:54321/graphql/v1
          DB URL: postgresql://postgres:postgres@localhost:54322/postgres
      Studio URL: http://localhost:54323
    Inbucket URL: http://localhost:54324
      JWT secret: super-secret-jwt-token-with-at-least-32-characters-long
        anon key: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0
service_role key: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0.EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU
```
``````

To add Supabase credentials to Astro, create an `.env` file in the root of the project with the following:

> ```ini title=".env"
> SUPABASE_URL=SUPABASE_URL
> SUPABASE_ANON_KEY=SUPABASE_ANON_KEY
> SUPABASE_JWT_SECRET=SUPABASE_JWT_SECRET
> ```

Reviewer 2: ``````suggestion
```ini
# .env
SUPABASE_URL=SUPABASE_URL
SUPABASE_ANON_KEY=SUPABASE_ANON_KEY
SUPABASE_JWT_SECRET=SUPABASE_JWT_SECRET
```
``````

> The Project structure should include this file:
> <FileTree title="Project Structure">
> - ...
> - supabase/
> - .env
> - ...
> </FileTree>

Reviewer 2: ```suggestion
The Project structure should now include this file:

<FileTree title="Project Structure">
- ...
- supabase/
- .env
- ...
</FileTree>
```

> 
> Now, these environment variables are available for use in the project.

Reviewer 2: ```suggestion
```

> To have IntelliSense for the Supabase environment variables, edit or create the file `env.d.ts` in `src/` directory and configure the types:

Reviewer 2: ```suggestion
To enable IntelliSense for the Supabase environment variables, edit or create the file `env.d.ts` in the`src/` directory and configure the types:
```

> ```ts title="src/env.d.ts"
> interface ImportMetaEnv {
>   readonly SUPABASE_URL: string;
>   readonly SUPABASE_ANON_KEY: string;
>   readonly SUPABASE_JWT_SECRET: string;
> }
> 
> interface ImportMeta {
>   readonly env: ImportMetaEnv;
> }
> ```

Reviewer 2: ``````suggestion
```ts
// src/env.d.ts
interface ImportMetaEnv {
  readonly SUPABASE_URL: string;
  readonly SUPABASE_ANON_KEY: string;
  readonly SUPABASE_JWT_SECRET: string;
}

interface ImportMeta {
  readonly env: ImportMetaEnv;
}
```
``````

:::tip
Read more about [environment variables](/en/guides/environment-variables/) and `.env` files in Astro.
:::

> The project structure should include these files:
> <FileTree title="Project Structure">
> - ...
> - supabase/
> - src/
>   - env.d.ts
> - .env
> - ...
> </FileTree>

Reviewer 2: ```suggestion
```

### Install Dependencies

To connect Astro with Supabase, install the following package using the command below:

> - `@supabase/supabase-js`
> 

Reviewer 2: ```suggestion
```

<PackageManagerTabs>
  <Fragment slot="npm">
    ```shell
    npm install @supabase/supabase-js
    ```
  </Fragment>
  <Fragment slot="pnpm">
    ```shell
    pnpm install @supabase/supabase-js
    ```
  </Fragment>
  <Fragment slot="yarn">
    ```shell
    yarn add @supabase/supabase-js
    ```
  </Fragment>
</PackageManagerTabs>

> ### Create Profile Table
> 1. Go to [Studio URL](http://localhost:54323)
> 2. Click on SQL Editor on the left sidebar
> 3. Choose `User Management Starter` from `Quick start` in the bottom of the page
> 4. Click `RUN`

Reviewer 2: ```suggestion
### Create Profile Table

1. Go to the Studio URL http://localhost:54323
2. Click on **SQL Editor** on the left sidebar
3. Choose **User Management Starter** from **Quick start** in the bottom of the page
4. Click **RUN**
```

### Create Supabase Client

1. Create a directory named `lib` in the `src/` directory.
2. Create a file named `supabaseClient.ts` in the `src/lib/` directory.

> In `supabaseClient.ts` add the following code to create the supabase client.

Reviewer 2: ```suggestion
In `supabaseClient.ts` add the following code to create the Supabase client:
```

> ```ts title="src/lib/supabaseClient.ts"
> import { createClient } from '@supabase/supabase-js'
> 
> export const supabase = createClient(
>   `${import.meta.env.SUPABASE_URL}`,
>   `${import.meta.env.SUPABASE_ANON_KEY}`,
>   {         
>     auth: {
>       persistSession: false,
>     }
>   }
> )
> ```

Reviewer 2: ``````suggestion
```ts 
// src/lib/supabaseClient.ts
import { createClient } from '@supabase/supabase-js'

export const supabase = createClient(
  `${import.meta.env.SUPABASE_URL}`,
  `${import.meta.env.SUPABASE_ANON_KEY}`,
  {         
    auth: {
      persistSession: false,
    }
  }
)
```
``````

> The project structure should include these new files:
> <FileTree title="Project Structure">
> - ...
> - src/
>   - lib/
>     - supabaseClient.ts
>   - env.d.ts
> - .env
> - astro.config.mjs
> - package.json
> - ...
> </FileTree>

Reviewer 2: ```suggestion

The project structure should include these new files:

<FileTree title="Project Structure">
- ...
- src/
  - lib/
    - supabaseClient.ts
  - env.d.ts
- .env
- astro.config.mjs
- package.json
- ...
</FileTree>
```

### Prepare Authentication Functions

- Create a file named `auth.ts` in the `src/lib/` directory.

In `auth.ts` add the following code to create the authentication functions:

> ```ts title="src/lib/auth.ts"
> import { supabase } from "./supabaseClient";
> 
> export async function signUp(email: string, password: string) {
>     return await supabase.auth.signUp({
>         email,
>         password,
>         options: {
>           emailRedirectTo: 'http://localhost:3000/profile'
>         }
>     })
> }
> 
> export async function signIn(email: string, password: string) {
>     return await supabase.auth.signInWithPassword({
>         email,
>         password
>     })
> }
> 
> export async function signInWithGithub() {
>     return await supabase.auth.signInWithOAuth({
>         provider: 'github',
>         options: {
>           redirectTo: 'http://localhost:3000/profile'
>         }
>     });
> }
> 
> export async function getSessionData() {
>     return await supabase.auth.getSession();
> }
> 
> export async function setSessionData(access_token: string, refresh_token: string) {
>     return await supabase.auth.setSession({
>         access_token,
>         refresh_token,
>     })
> }
> 
> export function AuthState() {
>     return supabase.auth.onAuthStateChange((event, session) => {
>         console.log(event, session);
>     })
> }
> 
> export async function signOut() {
>     return await supabase.auth.signOut()
> }
> ```

Reviewer 2: ``````suggestion
```ts
// src/lib/auth.ts
import { supabase } from "./supabaseClient";

export async function signUp(email: string, password: string) {
    return await supabase.auth.signUp({
        email,
        password,
        options: {
          emailRedirectTo: 'http://localhost:3000/profile'
        }
    })
}

export async function signIn(email: string, password: string) {
    return await supabase.auth.signInWithPassword({
        email,
        password
    })
}

export async function signInWithGithub() {
    return await supabase.auth.signInWithOAuth({
        provider: 'github',
        options: {
          redirectTo: 'http://localhost:3000/profile'
        }
    });
}

export async function getSessionData() {
    return await supabase.auth.getSession();
}

export async function setSessionData(access_token: string, refresh_token: string) {
    return await supabase.auth.setSession({
        access_token,
        refresh_token,
    })
}

export function AuthState() {
    return supabase.auth.onAuthStateChange((event, session) => {
        console.log(event, session);
    })
}

export async function signOut() {
    return await supabase.auth.signOut()
}
```
``````

:::tip
Check [Supabase User Management](https://supabase.com/dashboard/project/qiilhzulchhvmptaayfz/api?page=users) to learn more about these functions.
:::

> The project structure should include these new files:
> <FileTree title="Project Structure">

Reviewer 2: ```suggestion
The project structure should include these new files:

<FileTree title="Project Structure">
```

- ...
- src/
  - lib/
    - auth.ts
    - supabaseClient.ts
  - env.d.ts
- .env
- astro.config.mjs
- package.json
- ...
</FileTree>

### Sign Up

1. Go to `src/pages`
2. Create a new file named `signup.astro`
3. Add the following code to `signup.astro`:

```astro title="src/pages/signup.astro"
---
import Layout from "../layouts/Layout.astro";
import { getSessionData, setSessionData } from "../lib/auth"

let accessToken = Astro.cookies.get("sb-access-token").value
let refreshToken = Astro.cookies.get("sb-refresh-token").value

if (accessToken && refreshToken) {
    await setSessionData(accessToken, refreshToken)
    return Astro.redirect("/profile")
}

console.log(await getSessionData())
---

<Layout title="Sign Up">
    <div class="flex justify-center items-center">
        <div class="col-6 form-widget" aria-live="polite">
            <h1 class="text-3xl text-gray-800">Sign Up</h1>
            <form
                class="flex flex-col space-y-4 mt-8"
                aria-live="polite"
            >
                <input
                    class="border rounded px-3 py-2"
                    aria-live="polite"
                    name="email"

>                     type="email"
>                     placeholder="Email"
>                 />
>                 <input
>                     class="border rounded px-3 py-2"
>                     aria-live="polite"
>                     name="password"
>                     type="password"
>                     placeholder="Password"
>                 />
>                 <input
>                     class="border rounded px-3 py-2"
>                     aria-live="polite"
>                     name="confirm-password"
>                     type="password"
>                     placeholder="Confirm Password"
>                 />
>                 <button
>                     class="block px-3 py-2 w-full text-white bg-gray-800 border rounded hover:border-2"
>                     aria-live="polite"
>                     type="submit"
>                 >
>                     Sign Up
>                 </button>
>                 <p class="text-center text-gray-800">Already have an account? <a href="/signin">Sign In</a></p>
>             </form>
>         </div>
>     </div>        
> </Layout>
> 
> <script>
>     import { signUp } from "../lib/auth";
> 
>     const form = document.querySelector("form")
> 
>     form?.addEventListener("submit", async (e) => {
>         e.preventDefault();
>         console.log("submitted")
>         const formData = new FormData(e.target as HTMLFormElement);
>         const creds = {
>             email: formData.get("email") as string,
>             password: formData.get("password") as string,
>             confirmPassword: formData.get("confirm-password") as string,
>         };
>         
>         if (creds.password !== creds.confirmPassword) {
>             console.log("Passwords do not match");
>             alert("Passwords do not match");
>             return;
>         } else {
>             try {
>                 await signUp(creds.email, creds.password);
>             } catch (error) {
>                 console.log(error);
>             }
>         }
>     });
> </script>
> ```
> 

Reviewer 2: ``````suggestion
```astro
---
// src/pages/signup.astro
import Layout from "../layouts/Layout.astro";
import { getSessionData, setSessionData } from "../lib/auth"

let accessToken = Astro.cookies.get("sb-access-token").value
let refreshToken = Astro.cookies.get("sb-refresh-token").value

if (accessToken && refreshToken) {
    await setSessionData(accessToken, refreshToken)
    return Astro.redirect("/profile")
}

console.log(await getSessionData())
---

<Layout title="Sign Up">
    <div class="flex justify-center items-center">
        <div class="col-6 form-widget" aria-live="polite">
            <h1 class="text-3xl text-gray-800">Sign Up</h1>
            <form
                class="flex flex-col space-y-4 mt-8"
                aria-live="polite"
            >
                <input
                    class="border rounded px-3 py-2"
                    aria-live="polite"
                    name="email"
                    type="email"
                    placeholder="Email"
                />
                <input
                    class="border rounded px-3 py-2"
                    aria-live="polite"
                    name="password"
                    type="password"
                    placeholder="Password"
                />
                <input
                    class="border rounded px-3 py-2"
                    aria-live="polite"
                    name="confirm-password"
                    type="password"
                    placeholder="Confirm Password"
                />
                <button
                    class="block px-3 py-2 w-full text-white bg-gray-800 border rounded hover:border-2"
                    aria-live="polite"
                    type="submit"
                >
                    Sign Up
                </button>
                <p class="text-center text-gray-800">Already have an account? <a href="/signin">Sign In</a></p>
            </form>
        </div>
    </div>        
</Layout>

<script>
    import { signUp } from "../lib/auth";

    const form = document.querySelector("form")

    form?.addEventListener("submit", async (e) => {
        e.preventDefault();
        console.log("submitted")
        const formData = new FormData(e.target as HTMLFormElement);
        const creds = {
            email: formData.get("email") as string,
            password: formData.get("password") as string,
            confirmPassword: formData.get("confirm-password") as string,
        };
        
        if (creds.password !== creds.confirmPassword) {
            console.log("Passwords do not match");
            alert("Passwords do not match");
            return;
        } else {
            try {
                await signUp(creds.email, creds.password);
            } catch (error) {
                console.log(error);
            }
        }
    });
</script>
```

``````

Reviewer: following up the suggestion we should also remove styles and console logs to clear up the markup.

> :::note
> After signing up a user, Check [Inbucket URL](http://localhost:54324) to confirm email.
> :::

Reviewer 2: ```suggestion
:::note
After signing up a user, Check the Inbucket URL (http://localhost:54324) to confirm the email.
:::
```

> ### Sign In
> #### Sign In with Email

Reviewer 2: ```suggestion
### Sign In

#### Sign In with Email
```

1. Go to `src/pages`
2. Create a new file named `signin.astro`
3. Add the following code to `signin.astro`:

```astro title="src/pages/signin.astro"
---
import Layout from "../layouts/Layout.astro"
import { getSessionData, setSessionData } from "../lib/auth"

let accessToken = Astro.cookies.get("sb-access-token").value
let refreshToken = Astro.cookies.get("sb-refresh-token").value

if (accessToken && refreshToken) {
    await setSessionData(accessToken, refreshToken)
    return Astro.redirect("/profile")
}

console.log(await getSessionData())
---

<Layout title="Sign In">
    <div class="flex justify-center items-center">
        <div class="col-6 form-widget" aria-live="polite">
            <h1 class="text-3xl text-gray-800">Sign In</h1>
            <form
                class="flex flex-col space-y-4 mt-8"
                aria-live="polite"
            >

>                 <input
>                     class="border rounded px-3 py-2"
>                     aria-live="polite"
>                     name="email"
>                     type="email"
>                     placeholder="Email"
>                 />
>                 <input
>                     class="border rounded px-3 py-2"
>                     aria-live="polite"
>                     name="password"
>                     type="password"
>                     placeholder="Password"
>                 />
>                 <button
>                     class="block px-3 py-2 w-full text-white bg-gray-800 border rounded hover:border-2"
>                     aria-live="polite"
>                     type="submit"
>                 >
>                     Sign In
>                 </button>
>                 <p class="text-center text-gray-800">Need an account? <a href="/signup">Sign Up</a></p>
>             </form>
>         </div>
>     </div>        
> </Layout>
> 
> <script>
>     import { signIn, getSessionData } from "../lib/auth";
> 
>     const form = document.querySelector("form");
> 
>     form?.addEventListener("submit", async (e) => {
>         e.preventDefault();
>         const formData = new FormData(e.target as HTMLFormElement);
>         const creds = {
>             email: formData.get("email") as string,
>             password: formData.get("password") as string,
>         };
> 
>         const signin = await signIn(creds.email, creds.password);
>         if (signin.error !== null) {
>             console.log(signin.error);
>             alert(signin.error.message);
>             window.location.href = "/signin";
>         } else {
>             const session = await getSessionData();
> 
>             console.log(session);
> 
>             const maxAge = 15 * 24 * 60 * 60;
>             if(session) {
>                 document.cookie = `sb-access-token=${session.data.session?.access_token}; path=/; expires=${session.data.session?.expires_at}; max-age=${maxAge}; SameSite=Lax; secure`;
>                 document.cookie = `sb-refresh-token=${session.data.session?.refresh_token}; path=/; expires=${session.data.session?.expires_at}; max-age=${maxAge}; SameSite=Lax; secure`;
>                 window.location.href = "/profile";
>             }
>         }        
>     })
> </script>
> ```

Reviewer 2: ``````suggestion
```astro
---
// src/pages/signin.astro
import Layout from "../layouts/Layout.astro"
import { getSessionData, setSessionData } from "../lib/auth"

let accessToken = Astro.cookies.get("sb-access-token").value
let refreshToken = Astro.cookies.get("sb-refresh-token").value

if (accessToken && refreshToken) {
    await setSessionData(accessToken, refreshToken)
    return Astro.redirect("/profile")
}

console.log(await getSessionData())
---

<Layout title="Sign In">
    <div class="flex justify-center items-center">
        <div class="col-6 form-widget" aria-live="polite">
            <h1 class="text-3xl text-gray-800">Sign In</h1>
            <form
                class="flex flex-col space-y-4 mt-8"
                aria-live="polite"
            >
                <input
                    class="border rounded px-3 py-2"
                    aria-live="polite"
                    name="email"
                    type="email"
                    placeholder="Email"
                />
                <input
                    class="border rounded px-3 py-2"
                    aria-live="polite"
                    name="password"
                    type="password"
                    placeholder="Password"
                />
                <button
                    class="block px-3 py-2 w-full text-white bg-gray-800 border rounded hover:border-2"
                    aria-live="polite"
                    type="submit"
                >
                    Sign In
                </button>
                <p class="text-center text-gray-800">Need an account? <a href="/signup">Sign Up</a></p>
            </form>
        </div>
    </div>        
</Layout>

<script>
    import { signIn, getSessionData } from "../lib/auth";

    const form = document.querySelector("form");

    form?.addEventListener("submit", async (e) => {
        e.preventDefault();
        const formData = new FormData(e.target as HTMLFormElement);
        const creds = {
            email: formData.get("email") as string,
            password: formData.get("password") as string,
        };

        const signin = await signIn(creds.email, creds.password);
        if (signin.error !== null) {
            console.log(signin.error);
            alert(signin.error.message);
            window.location.href = "/signin";
        } else {
            const session = await getSessionData();

            console.log(session);

            const maxAge = 15 * 24 * 60 * 60;
            if(session) {
                document.cookie = `sb-access-token=${session.data.session?.access_token}; path=/; expires=${session.data.session?.expires_at}; max-age=${maxAge}; SameSite=Lax; secure`;
                document.cookie = `sb-refresh-token=${session.data.session?.refresh_token}; path=/; expires=${session.data.session?.expires_at}; max-age=${maxAge}; SameSite=Lax; secure`;
                window.location.href = "/profile";
            }
        }        
    })
</script>
```
``````

    import { signUp, signInWithGithub } from "../lib/auth";

    const form = document.querySelector("form")

    form?.addEventListener("submit", async (e) => {
        e.preventDefault();
        console.log("submitted")
        const formData = new FormData(e.target as HTMLFormElement);
        const creds = {
            email: formData.get("email") as string,
            password: formData.get("password") as string,
            confirmPassword: formData.get("confirm-password") as string,
        };
        
        if (creds.password !== creds.confirmPassword) {
            console.log("Passwords do not match");
            alert("Passwords do not match");
            return;
        } else {
            try {
                await signUp(creds.email, creds.password);
            } catch (error) {
                console.log(error);
            }
        }
    });

    const github = document.getElementById("github");
    github?.addEventListener("click", async (e) => {
        e.preventDefault();
        await signInWithGithub();
    })
</script>
```

### Profile

1. Go to `src/pages`
2. Create a new file named `profile.astro`
3. Add the following code to `profile.astro`:

> ```astro title="src/pages/profile.astro"
> ---
> import Layout from "../layouts/Layout.astro";
> import { getSessionData, setSessionData, AuthState} from "../lib/auth";
> 
> let accessToken = Astro.cookies.get("sb-access-token").value
> let refreshToken = Astro.cookies.get("sb-refresh-token").value
> 
> console.log("accessToken", accessToken);
> console.log("refreshToken", refreshToken);
> const session = await setSessionData(accessToken, refreshToken);
> 
> console.log("session", session);
> 
> if (session.data.session === null || (accessToken === undefined && refreshToken === undefined)) {
> 	return Astro.redirect("/signin");
> }
> 
> console.log("Auth State", AuthState())
> ---
> 
> <Layout title="Profile">
> 	<div class="flex flex-col items-center justify-center">
> 		<h1 class="text-2xl font-bold text-gray-800">User Data</h1>
> 		<p class="text-gray-800">{JSON.stringify(session)}</p>
>     </div>
> 	<button
> 		class="block px-3 py-2 w-full text-white bg-gray-800 border rounded hover:border-2"
> 		type="button"
>         id="sign-out-button"
> 	>
> 		Sign Out
> 	</button>
> </Layout>
> 
> <script>
>     const signOutButton = document.getElementById("sign-out-button");
>     signOutButton.addEventListener("click", async () => {
> 		const response = await fetch("/api/signout")
> 
> 		const data = await response.json();
> 
> 		console.log(data);
> 
> 		if (data.message === "success") {
> 			window.location.href = "/signin";
> 		}
>     });
> </script>
> ```

Reviewer 2: ``````suggestion
```astro
---
// src/pages/profile.astro
import Layout from "../layouts/Layout.astro";
import { getSessionData, setSessionData, AuthState} from "../lib/auth";

let accessToken = Astro.cookies.get("sb-access-token").value
let refreshToken = Astro.cookies.get("sb-refresh-token").value

console.log("accessToken", accessToken);
console.log("refreshToken", refreshToken);
const session = await setSessionData(accessToken, refreshToken);

console.log("session", session);

if (session.data.session === null || (accessToken === undefined && refreshToken === undefined)) {
	return Astro.redirect("/signin");
}

console.log("Auth State", AuthState())
---

<Layout title="Profile">
	<div class="flex flex-col items-center justify-center">
		<h1 class="text-2xl font-bold text-gray-800">User Data</h1>
		<p class="text-gray-800">{JSON.stringify(session)}</p>
    </div>
	<button
		class="block px-3 py-2 w-full text-white bg-gray-800 border rounded hover:border-2"
		type="button"
        id="sign-out-button"
	>
		Sign Out
	</button>
</Layout>

<script>
    const signOutButton = document.getElementById("sign-out-button");
    signOutButton.addEventListener("click", async () => {
		const response = await fetch("/api/signout")

		const data = await response.json();

		console.log(data);

		if (data.message === "success") {
			window.location.href = "/signin";
		}
    });
</script>
```
``````

> The project structure should include these new files:
> <FileTree title="Project Structure">
> - ...
> - supabase/
>   - config.toml
>   - functions/
>   - migrations/
>   - seed.sql
> - src/
>   - pages/
>     - signup.astro
>     - signin.astro
>     - profile.astro
>   - lib/
>     - auth.ts
>     - supabaseClient.ts
>   - env.d.ts
> - .env
> - astro.config.mjs
> - package.json
> - ...
> </FileTree>

Reviewer 2: ```suggestion
```

### Sign Out Endpoint

1. Go to `src/pages/`
2. Create a new directory named `api/`
3. Create a new file named `signout.ts`

> Add the following code to `signout.ts`:
> ```ts title="src/pages/api/signout.ts"
> import type { APIRoute } from "astro";
> import { signOut } from "../../lib/auth";
> 
> export const get: APIRoute = async ({ cookies }) => {
>     cookies.delete("sb-access-token", { path: "/" });
>     cookies.delete("sb-refresh-token", { path: "/" });
>     try {
>         const { error } = await signOut()
> 
>         if (error) {
>             console.log("error", error.message);
>         } else {
>             return new Response(
>                 JSON.stringify({ message: "success" }),
>                 { status: 200 }
>             )
>         }
>     } catch (error) {
>         console.log("error", error);
>         return new Response(
>             JSON.stringify({ error: error.message }),
>             { status: 500 }
>         )
>     }
> }
> ```

Reviewer 2: ``````suggestion
Add the following code to `signout.ts`:

```ts
// src/pages/api/signout.ts
import type { APIRoute } from "astro";
import { signOut } from "../../lib/auth";

export const get: APIRoute = async ({ cookies }) => {
    cookies.delete("sb-access-token", { path: "/" });
    cookies.delete("sb-refresh-token", { path: "/" });
    try {
        const { error } = await signOut()

        if (error) {
            console.log("error", error.message);
        } else {
            return new Response(
                JSON.stringify({ message: "success" }),
                { status: 200 }
            )
        }
    } catch (error) {
        console.log("error", error);
        return new Response(
            JSON.stringify({ error: error.message }),
            { status: 500 }
        )
    }
}
```
``````

> The project structure should include these new files:
> <FileTree title="Project Structure">
> - ...
> - supabase/
>   - config.toml
>   - functions/
>   - migrations/
>   - seed.sql
> - src/
>   - pages/
>     - api/
>       - signout.ts
>     - signup.astro
>     - signin.astro
>     - profile.astro
>   - lib/
>     - auth.ts
>     - supabaseClient.ts
>   - env.d.ts
> - .env
> - astro.config.mjs
> - package.json
> - ...
> </FileTree>

Reviewer 2: ```suggestion

Finally, the project structure will look like this:

<FileTree title="Project Structure">
- ...
- supabase/
  - config.toml
  - functions/
  - migrations/
  - seed.sql
- src/
  - pages/
    - api/
      - signout.ts
    - signup.astro
    - signin.astro
    - profile.astro
  - lib/
    - auth.ts
    - supabaseClient.ts
  - env.d.ts
- .env
- astro.config.mjs
- package.json
- ...
</FileTree>
```

description: Add a backend to your project with Supabase
type: backend
service: Supabase
---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import FileTree from '~/components/FileTree.astro'

[Supabase](https://supabase.com/) is an open source Firebase alternative. It provides a Postgres database, authentication, edge functions, realtime subscriptions, and storage.

## Initializing Supabase in Astro

### Prerequisites

- A Supabase project. If you don't have one, you can sign up for free at [supabase.com](https://supabase.com/) and create a new project.

> - An Astro project with [server-side rendering (SSR)](/en/guides/server-side-rendering/#enabling-ssr-in-your-project) enabled.

Reviewer 3: ```suggestion
- An Astro project with [server-side rendering (SSR)](/en/guides/server-side-rendering/) enabled.
```
Just future-proofing an anchor link I know will change!

- Supabase credentials for your project. You can find these in the **Settings > API** tab of your Supabase project.
  - `SUPABASE_URL`: The URL of your Supabase project.
  - `SUPABASE_ANON_KEY`: The anonymous key for your Supabase project.

### Adding Supabase credentials

To add your Supabase credentials to your Astro project, add the following to your `.env` file:

```ini title=".env"
SUPABASE_URL=YOUR_SUPABASE_URL
SUPABASE_ANON_KEY=YOUR_SUPABASE_ANON_KEY
```

Now, these environment variables are available in your project.

If you would like to have IntelliSense for your environment variables, edit or create the `env.d.ts` in your `src/` directory and add the following:

```ts title="src/env.d.ts"
interface ImportMetaEnv {
  readonly SUPABASE_URL: string
  readonly SUPABASE_ANON_KEY: string
}

interface ImportMeta {
  readonly env: ImportMetaEnv
}
```

:::tip
Read more about [environment variables](/en/guides/environment-variables/) and `.env` files in Astro.
:::

> Your project should now include these new files:

Reviewer 3: ```suggestion
Your project should now include these files:
```
Just in case they weren't new!
